### PR TITLE
Add Service Ownership Flow

### PR DESCRIPTION
**Pull request by Yasirmod17**
## Summary

Extends ownership for the service element to allow more refined ownership concept

## How did you test this change?

- [x] Test(s) added
- [ ] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

after:

1. the old flow is kept 

https://github.com/finos/legend-studio/assets/73408381/8e9ef49a-666f-4c76-9be7-b36e4a983059

2. 

https://github.com/finos/legend-studio/assets/73408381/4e759162-f73d-4139-9c0d-23843984bb66

### CODE REVIEW

File: packages/legend-application-studio/src/components/editor/editor-group/service-editor/ServiceEditor.tsx [hunk 1]
```diff
@@ -142,6 +152,62 @@ const ServiceGeneralEditor = observer(() => {
...
+      ) {
+        service_updateUserOwnership(ownership, ownerInputValue, idx);
+      }
+    };
+
+  const deleteUser =
+    (idx: number): (() => void) =>
+    (): void => {
+      if (!isReadOnly && ownership instanceof UserListOwnership) {
+        console.log('deleting');
```
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/components/editor/editor-group/service-editor/ServiceEditor.tsx:199]: :`) miss this

File: packages/legend-graph/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_ServiceTransformer.ts [hunk 2]
```diff
@@ -183,6 +193,24 @@ const transformSingleExecution = (
...
 
+const transformDeploymentOwnership = (
+  element: DeploymentOwnership,
+  context: V1_GraphTransformerContext,
+): V1_DeploymentOwnership => {
+  const ownership = new V1_DeploymentOwnership();
+  ownership.identifier = element.identifier;
+  return ownership;
+};
+
```
Comment (YannanGao-gs) [inline packages/legend-graph/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_ServiceTransformer.ts:204]: is it possible maybe we could move these down a little bit to be closer to transformServiceOwnership()

File: packages/legend-graph/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_ServiceTransformer.ts [hunk 3]
```diff
@@ -224,6 +252,21 @@ const transformServiceExecution = (
   );
 };
 
+const transformServiceOwnership = (
+  metamodel: ServiceOwnership,
+  context: V1_GraphTransformerContext,
+): V1_ServiceOwnership => {
+  if (metamodel instanceof DeploymentOwnership) {
```
Comment (YannanGao-gs) [inline packages/legend-graph/src/graph-manager/protocol/pure/v1/transformation/pureGraph/from/V1_ServiceTransformer.ts:259]: context could be removed. it's not used

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts [hunk 4]
```diff
@@ -43,11 +43,15 @@ import {
   type Type,
   resolveServiceQueryRawLambda,
   PureExecution,
+  DeploymentOwnership,
+  ServiceOwnership,
```
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:47]: type ServiceOwnership
could we also move it to the top of this section pls?

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts [hunk 5]
```diff
@@ -101,6 +110,22 @@ export const isServiceQueryTDS = (
...
 
 export const MINIMUM_SERVICE_OWNERS = 2;
+export const OWNERSHIP_OPTIONS = [
+  {
+    label: 'Deployment',
+    value: ServiceOwnershipType.DEPLOYMENT_OWNERSHIP,
+  },
+  {
+    label: 'User List',
+    value: ServiceOwnershipType.USERLIST_OWNERSHIP, //"UserListOwnership",
```
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:120]: :') maybe we could delete the comment?

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts [hunk 6]
```diff
@@ -101,6 +110,22 @@ export const isServiceQueryTDS = (
...
+  },
+];
+
+export type ServiceOwnerOption = {
```
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:124]: maybe just to be consistent. is it better to move it to ServiceEditor.tsx under line 
`type UserOption = { label: string; value: string };`

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts [hunk 7]
```diff
@@ -153,6 +180,41 @@ export class ServiceEditorState extends ElementEditorState {
...
+    } else if (ownership instanceof UserListOwnership) {
+      return {
+        label: 'UserList',
+        value: ServiceOwnershipType.USERLIST_OWNERSHIP,
+      };
+    }
+    return undefined;
+  }
+
+  setSelectedOwnership(o: { label: string; value: string }): void {
```
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:199]: :ServiceOwnerOption

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts [hunk 8]
```diff
@@ -153,6 +180,41 @@ export class ServiceEditorState extends ElementEditorState {
...
+    let newVal: ServiceOwnership | undefined;
+    switch (o.value) {
+      case ServiceOwnershipType.DEPLOYMENT_OWNERSHIP: {
+        newVal = new DeploymentOwnership('', this.service);
+        service_setOwnership(this.service, newVal);
```
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:204]: i prefer removing newVal

File: packages/legend-graph/src/graph-manager/action/changeDetection/DSL_Service_ObserverHelper.ts [hunk 9]
```diff
@@ -224,6 +229,37 @@ export const observe_MultiExecutionTest = skipObserved(
   },
 );
 
+export const observe_Ownership = (
```
Comment (YannanGao-gs) [inline packages/legend-graph/src/graph-manager/action/changeDetection/DSL_Service_ObserverHelper.ts:232]: this function needs to be defined after observe_deploymentOwnership and observe_userListOwnership

File: packages/legend-application-studio/src/stores/graph-modifier/DSL_Service_GraphModifierHelper.ts [hunk 10]
```diff
@@ -23,6 +23,9 @@ import {
   type RawLambda,
   type Runtime,
   type Service,
+  DeploymentOwnership,
+  UserListOwnership,
```
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/stores/graph-modifier/DSL_Service_GraphModifierHelper.ts:27]: type ServiceOwnership
type UserListOwnership

File: packages/legend-application-studio/src/components/editor/editor-group/service-editor/ServiceEditor.tsx [hunk 11]
```diff
@@ -285,157 +359,348 @@ const ServiceGeneralEditor = observer(() => {
...
-                        disabled={isReadOnly}
-                        value={ownerInputValue}
-                        onChange={changeOwnerInputValue}
-                      />
-                      <div className="panel__content__form__section__list__new-item__actions">
-                        <button
-                          className="panel__content__form__section__list__new-item__add-btn btn btn--dark"
-                          disabled={
-                            isReadOnly || owners.includes(ownerInputValue)
+        {owners.length == 0 && (
```
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/components/editor/editor-group/service-editor/ServiceEditor.tsx:362]: Screenshot 2023-09-18 at 3 38 54 PM
could we do something like this?
The space between 'Owners' and  'Auto Activate Updates' will be kept

Comment (YannanGao-gs) [reply packages/legend-application-studio/src/components/editor/editor-group/service-editor/ServiceEditor.tsx:362]: owners.length === 0

Comment (YannanGao-gs) [inline packages/legend-application-studio/src/components/editor/editor-group/service-editor/ServiceEditor.tsx:362]: Screenshot 2023-09-18 at 3 38 54 PM
could we wrap up the code for owners so the space between `owners` and `Activate` will be kept

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts (hunk 6 above)
Comment (Yasirmod17) [reply packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:124]: the type is used in a function in editor state too so can't move it to the editor

Comment (YannanGao-gs) [reply packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:124]: you could export it

File: .changeset/new-spiders-sparkle.md [hunk 12]
```diff
@@ -0,0 +1,6 @@
+---
+'@finos/legend-application-studio': patch
+'@finos/legend-graph': patch
+---
```
Comment (YannanGao-gs) [inline .changeset/new-spiders-sparkle.md:4]: could we separate '@finos/legend-application-studio' and  '@finos/legend-graph' into 2 changesets? 
Leave the description of one that contains '@finos/legend-graph' empty

Comment (MauricioUyaguari) [reply .changeset/new-spiders-sparkle.md:4]: the change in studio should also not be patach but minor. thanks

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts [hunk 13]
```diff
@@ -153,6 +180,43 @@ export class ServiceEditorState extends ElementEditorState {
     this.selectedTab = SERVICE_TAB.TEST;
   }
 
+  get selectedOwnership(): ServiceOwnerOption | undefined {
+    const ownership = this.service.ownership;
+    if (ownership instanceof DeploymentOwnership) {
+      return {
+        label: 'Deployment',
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:187]: lets move this string as const on top as we are repeating them. 

so like const DEPLOYMENT = 'deployment'

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts [hunk 14]
```diff
@@ -153,6 +180,43 @@ export class ServiceEditorState extends ElementEditorState {
...
+    switch (o.value) {
+      case ServiceOwnershipType.DEPLOYMENT_OWNERSHIP: {
+        service_setOwnership(
+          this.service,
+          new DeploymentOwnership('', this.service),
+        );
+        break;
+      }
+      case ServiceOwnershipType.USERLIST_OWNERSHIP: {
+        newVal = new UserListOwnership([], this.service);
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:210]: we previously added the current user as the first owner if they choose owner list. should we still do this when they switch to userlist ?

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts [hunk 15]
```diff
@@ -153,6 +180,43 @@ export class ServiceEditorState extends ElementEditorState {
...
+        service_setOwnership(this.service, newVal);
+        break;
+      }
+      default: {
+        throw new Error('Unsupported');
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:215]: we dont want to throw errors in the UI even if they are unlikely as that breaks the app. can we either break in the default case here or alert user their selection is not supported. See usage of `notifyError` in codebase.

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts (hunk 14 above)
Comment (YannanGao-gs) [inline packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts:210]: could we also remove this and line 200?

Comment (Yasirmod17) [reply packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts]: Changed

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts (hunk 15 above)
Comment (Yasirmod17) [reply packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts]: changed

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts (hunk 14 above)
Comment (Yasirmod17) [reply packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts]: done

File: packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts (hunk 13 above)
Comment (Yasirmod17) [reply packages/legend-application-studio/src/stores/editor/editor-state/element-editor-state/service/ServiceEditorState.ts]: done

File: .changeset/new-spiders-sparkle.md [hunk 16]
```diff
@@ -0,0 +1,6 @@
...
+'@finos/legend-application-studio': minor
+'@finos/legend-graph': minor
+---
```
Comment (YannanGao-gs) [inline .changeset/new-spiders-sparkle.md:4]: two changesets 🥹

**Review by MauricioUyaguari — APPROVED** [review]

**Pull request merged**